### PR TITLE
Add tracking of the distribution group ID to Analytics

### PR DESCRIPTION
**Pull request by roland-chernov-akvelon**
This PR:
- Adds `distributionGroupId` field to sent logs (if exists).
- Adds a channel listener in the Distribution module which sets the `distributionGroupId` value.

### CODE REVIEW

File: sdk/appcenter-distribute/src/debug/res/values/features.xml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources xmlns:tools="http://schemas.android.com/tools" tools:ignore="TypographyDashes">
+    <string name="appcenter_distribute_distribution_group_id_feature_enabled" tools:ignore="MissingTranslation">true</string>
```
Comment (MatkovIvan) [inline sdk/appcenter-distribute/src/debug/res/values/features.xml]: `translatable="false"`?

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 2]
```diff
@@ -759,6 +779,9 @@ private void decryptAndGetReleaseDetails(String updateToken, String distribution
         /* If the group was from Mobile Center storage, save it in the new storage. */
         if (mobileCenterFailOver) {
             PreferencesStorage.putString(PREFERENCE_KEY_DISTRIBUTION_GROUP_ID, distributionGroupId);
+            if (this.isDistributionGroupIdFeatureEnabled()) {
```
Comment (MatkovIvan) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Why not just `mDistributeInfoTracker != null` instead?

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java [hunk 3]
```diff
@@ -441,6 +455,12 @@ public void run() {
             PreferencesStorage.remove(PREFERENCE_KEY_POSTPONE_TIME);
             PreferencesStorage.remove(PREFERENCE_KEY_UPDATE_SETUP_FAILED_PACKAGE_HASH_KEY);
             PreferencesStorage.remove(PREFERENCE_KEY_UPDATE_SETUP_FAILED_MESSAGE_KEY);
+
+            /* Disable the distribute info tracker if the feature is enabled. */
+            if (this.isDistributionGroupIdFeatureEnabled() && mDistributeInfoTracker != null) {
```
Comment (MatkovIvan) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: Looks like checking `isDistributionGroupIdFeatureEnabled` is not necessary here.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        mDistributionGroupId = distributionGroupId;
+    }
+
+    @Override
+    public void onEnqueuingLog(@NonNull Log log, @NonNull String groupName) {
+        if (mDistributionGroupId == null) {
+            return;
+        }
+
+        /* Set current distribution group ID */
```
Comment (MatkovIvan) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java]: A dot at the end.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    /**
+     * Update the distribution group ID value that is added to logs.
+     *
+     * @param distributionGroupId the distribution group ID value that is added to logs.
+     */
+    public void updateDistributionGroupId(String distributionGroupId) {
+        mDistributionGroupId = distributionGroupId;
+    }
+
```
Comment (MatkovIvan) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java]: Extra empty line.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/ingestion/models/Log.java [hunk 6]
```diff
@@ -40,6 +40,20 @@
      */
     void setSid(UUID sid);
 
+    /**
+     * Get the distributionGroupId value.
+     *
+     * @return the distributionGroupId value
```
Comment (MatkovIvan) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/ingestion/models/Log.java]: A dot at the end.

File: sdk/appcenter/src/main/java/com/microsoft/appcenter/ingestion/models/AbstractLog.java [hunk 7]
```diff
@@ -81,6 +92,24 @@ public void setSid(UUID sid) {
         this.sid = sid;
     }
 
+    /**
+     * Get the distributionGroupId value.
+     *
+     * @return the distributionGroupId value
```
Comment (MatkovIvan) [inline sdk/appcenter/src/main/java/com/microsoft/appcenter/ingestion/models/AbstractLog.java]: A dot at the end.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java (hunk 3 above)
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/Distribute.java]: This PR is not user-facing so I've removed feature flag. This check removed as well. Thank you.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java (hunk 5 above)
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java]: Fixed. Thanks.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java (hunk 4 above)
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java]: Fixed. Thanks.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+     * Init.
+     *
+     * @param distributionGroupId distribution group ID that is added to logs (if exists).
+     */
+    public DistributeInfoTracker(String distributionGroupId) {
+        mDistributionGroupId = distributionGroupId;
+    }
+
+    @Override
+    public void onEnqueuingLog(@NonNull Log log, @NonNull String groupName) {
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java]: please add synchronized before public and void

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        /* Set current distribution group ID. */
+        log.setDistributionGroupId(mDistributionGroupId);
+    }
+
+    /**
+     * Update the distribution group ID value that is added to logs.
+     *
+     * @param distributionGroupId the distribution group ID value that is added to logs.
+     */
+    public void updateDistributionGroupId(String distributionGroupId) {
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java]: please add synchronized before public and void

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        mDistributionGroupId = distributionGroupId;
+    }
+
+    /**
+     * Don't add the distribution group ID value to logs.
+     */
+    public void removeDistributionGroupId() {
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java]: please add synchronized before public and void

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTrackerTest.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.UUID;
+
+import static junit.framework.Assert.assertNull;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Mockito.mock;
+
+
+@SuppressWarnings("unused")
+@PrepareForTest({DistributeInfoTracker.class})
```
Comment (guperrot) [inline sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTrackerTest.java]: You can remove curly braces when there is only 1, this what we do elsewhere, should use same code convention.

File: sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java (hunk 8 above)
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTracker.java]: Fixed. Thank you.

File: sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTrackerTest.java (hunk 11 above)
Comment (roland-chernov-akvelon) [reply sdk/appcenter-distribute/src/test/java/com/microsoft/appcenter/distribute/channel/DistributeInfoTrackerTest.java]: Got it. Thank you.

**Pull request merged**